pkg/istio/authorizationpolicy/controller.go

```diff
@@ -0,0 +1,463 @@
+		// to prevent user manually edit authorization policy files
+		if _, ok = obj.Annotations["overrideAuthzPolicy"]; !ok {
+			athenzDomainRaw, exists, err := c.AdIndexInformer.GetIndexer().GetByKey(athenz.NamespaceToDomain(obj.Namespace))
+			if err != nil {
+				return err
+			}
+
+			if !exists {
+				return fmt.Errorf("athenz domain %v does not exist in cache", obj.Namespace)
+			}
+
+			athenzDomain, ok := athenzDomainRaw.(*adv1.AthenzDomain)
+			if !ok {
+				return errors.New("athenz domain cast failed")
+			}
+			signedDomain := athenzDomain.Spec.SignedDomain
+			// regenerate authz policy spec, since for authz policy's name match with service's label 'app' value
+			// it can just pass in authz policy name as arg to func convertAthenzModelIntoIstioAuthzPolicy
+			label := obj.Spec.Selector.MatchLabels["svc"]
+			domainRBAC := m.ConvertAthenzPoliciesIntoRbacModel(signedDomain.Domain, &c.AdIndexInformer)
+			convertedCR := c.RbacProviderV2.ConvertAthenzModelIntoIstioAuthzPolicy(domainRBAC, obj.Namespace, obj.Name, label)
+			if !c.Dryrun {
+				// prevent manual editing the file
+				if castItem.Operation == model.EventUpdate {
+					// assign current revision, update function requires a defined resource version
+					convertedCR.ResourceVersion = obj.ResourceVersion
+					_, err := c.ConfigStoreCache.Update(convertedCR)
+					if err != nil {
+						log.Errorln("error updating authz policy: ", err.Error())
+						return err
+					}
+				}
+				// prevent
+				if castItem.Operation == model.EventDelete {
+					_, err := c.ConfigStoreCache.Create(convertedCR)
+					if err != nil {
+						log.Errorln("error updating authz policy: ", err.Error())
+						return err
+					}
+				}
+			} else {
+				err := c.createDryrunResource(convertedCR, obj.Name, obj.Namespace)
+				if err != nil {
+					return fmt.Errorf("unable write to file, err: %v", err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// checkAuthzEnabledAnnotation checks if current servce object has "authz.istio.io/enabled" annotation set
+func (c *Controller) checkAuthzEnabledAnnotation(serviceObj *corev1.Service) bool {
+	if _, ok := serviceObj.Annotations[authzEnabledAnnotation]; ok {
+		if serviceObj.Annotations[authzEnabledAnnotation] == authzEnabled {
+			return true
+		} else {
+			return false
+		}
+	}
+	return false
+}
+
+func (c *Controller) createDryrunResource(convertedCR model.Config, authzPolicyName string, namespace string) error {
+	convertedObj, err := crd.ConvertConfig(collections.IstioSecurityV1Beta1Authorizationpolicies, convertedCR)
+	if err != nil {
+		log.Errorln("Unable to convert authorization policy config to istio objects")
+	}
+	configInBytes, err := yaml.Marshal(convertedObj)
+	if err != nil {
+		return fmt.Errorf("could not marshal %v: %v", convertedCR.Name, err)
+	}
+	yamlFileName := authzPolicyName + "--" + namespace + ".yaml"
+	err = ioutil.WriteFile(DryRunStoredFilesDirectory + yamlFileName, configInBytes, 0644)
```

Comment:
Should we directly return with the WriteFile call? We have the log line with the "unable write to file" string everywhere we call this function and it returns an error.